File: dotnet/Controllers/RedisResearcherController.cs
using BoilerplateCombo.Models;
using BoilerplateCombo.Service;
using Microsoft.AspNetCore.Mvc;

namespace BoilerplateCombo.Controllers;

[ApiController]
[Route("redis/researcher")]
public class RedisResearcherController : ControllerBase
{
    private readonly RedisResearcherService _service;

    public RedisResearcherController(RedisResearcherService service)
    {
        _service = service;
    }

    [HttpGet]
    public async Task<IActionResult> GetResearchers(int page = 1, int pageSize = 25, string sortBy = "id", bool ascending = true, string? filter = null)
    {
        var (researchers, totalCount) = await _service.GetResearchersAsync(page, pageSize, sortBy, ascending, filter);
        return Ok(new { data = researchers, totalCount });
    }

    [HttpGet("export")]
    public async Task<IActionResult> ExportResearchers()
    {
        var csv = await _service.ExportResearchersAsCsvAsync();
        return File(System.Text.Encoding.UTF8.GetBytes(csv), "text/csv", "researchers.csv");
    }

    [HttpPost]
    public async Task<IActionResult> AddResearcher([FromBody] Researcher2 researcher)
    {
        var added = await _service.AddResearcherAsync(researcher);
        return CreatedAtAction(nameof(GetResearchers), new { id = added.id }, added);
    }

    [HttpPut("{id}")]
    public async Task<IActionResult> UpdateResearcher(int id, [FromBody] Researcher2 researcher)
    {
        var updated = await _service.UpdateResearcherAsync(id, researcher);
        if (!updated)
            return NotFound();
        return NoContent();
    }

    [HttpDelete("{id}")]
    public async Task<IActionResult> DeleteResearcher(int id)
    {
        var deleted = await _service.DeleteResearcherAsync(id);
        if (!deleted)
            return NotFound();
        return NoContent();
    }
}

File: springboot/src/main/java/com/boilerplatecombo/controller/RedisResearcherController.java
package com.boilerplatecombo.controller;

import com.boilerplatecombo.entity.Researcher2;
import com.boilerplatecombo.service.RedisResearcherService;
import org.springframework.context.annotation.Lazy;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;
import java.util.concurrent.CompletableFuture;

@RestController
@RequestMapping("/redis/researcher")
public class RedisResearcherController {

    private RedisResearcherService service;

    public RedisResearcherController(@Lazy RedisResearcherService researcherService) {
        this.service = researcherService;
    }

    @GetMapping
    public ResponseEntity<CompletableFuture<Map<String, String>>> getResearchers(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "25") int pageSize,
            @RequestParam(defaultValue = "id") String sortBy,
            @RequestParam(defaultValue = "true") boolean ascending,
            @RequestParam(required = false) String filter
    ) {
        return ResponseEntity.ok(service.getResearcherByIdAsync(1));
    }

    @PostMapping
    public ResponseEntity<String> addResearcher(@RequestBody Researcher2 researcher) {
        service.addResearcherAsync(researcher);
        return ResponseEntity.status(201).body("");
    }

    @PutMapping("/{id}")
    public ResponseEntity<String> updateResearcher(@PathVariable int id, @RequestBody Researcher2 researcher) {
        CompletableFuture<Long> updated = service.updateResearcherAsync(id, researcher);
        return ResponseEntity.status(200).body("");
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<String> deleteResearcher(@PathVariable int id) {
        CompletableFuture<Boolean> deleted = service.deleteResearcherAsync(id);
        return ResponseEntity.status(200).body("");
    }
}
